### PR TITLE
Enable kernel modules

### PR DESCRIPTION
**Pull request by zx2c4**
For years, I was in favor of disabling module loading, so that I could produce my ultra minimal custom-tailored kernel only containing exactly what I needed and nothing more. Then Linux became complicated, and it became clear this was no longer a good strategy.

There are too many random modules that do various things -- netfilter matches, input device drivers, and so forth -- that you simply _do not want to always load_. They not only take up memory, but they use slots in various linked lists and hash tables, making traversal that much slower.

And more importantly, many modules just aren't that well written. While it's nice to have the choice in case I need it, most of the time I don't want some of the garbage loaded inside my kernel.

In other words, while the aesthetic used to be geared toward keeping trash out of your build, the new cleanliness is more about keeping trash out of your TCB.

There's another reason why I like this, too. I'm the developer of [WireGuard](https://www.wireguard.io/), a next generation high performance Linux kernel VPN. I plan to produce modules for the Oneplus 3T, most likely against your kernel. Obviously having modules enabled is a prerequisite for that.

### CODE REVIEW

**Comment by kerneltoast** [comment]
Kernel modules are great, but Android doesn't have a daemon to automatically load modules. Without kmod, all modules must be manually loaded by init, either unconditionally or upon a system property change. Android kernels are machine specific and very stripped down, so the overhead from building in all of the included drivers is minimal.

If this kernel were meant to run on multiple machines with a module-friendly OS, then modules would definitely be the way the go. However, Android just isn't set up for this.

**Comment by zx2c4** [comment]
> Kernel modules are great, but Android doesn't have a daemon to automatically load modules.

All of the drivers that you'd want to be compiled as modules will autoload via the call to `request_module` -> `call_modprobe`:
https://git.zx2c4.com/linux/tree/kernel/kmod.c#n69
Which simply invokes `modprobe` in userspace.

The drivers that don't get loaded via `request_module` are better off compiled in.

> Android kernels are machine specific and very stripped down, so the overhead from building in all of the included drivers is minimal.

I think not. Look at how many different networking protocols and netfilter connection tracking classifiers you have enabled in this kernel. It's a bloated mess. Hell, there's even an Xbox driver compiled in. This is *not* a minimal kernel build, by any measure. Probably most of the networking subsystem could be modularized, inserted automatically via `request_module`, which alone would be a huge improvement. Anything not inserted automatically that way can be left compiled in.

> Android just isn't set up for this.

This just isn't true. Most Android kernels come with module loading enabled, including LineageOS's Oneplus kernels. On the contrary, it's more of an odd quirk that your kernel doesn't have modules enabled.

**Comment by kerneltoast** [comment]
Android userspace doesn't have modprobe; it only has insmod. This is how the OEM uses modules:
Unconditionally: https://github.com/OnePlusOSS/android_vendor_qcom_copper/blob/oneplusX/6.0.1/init.target.rc#L72
On a system property change:
https://github.com/OnePlusOSS/android_device_qcom_common/blob/oneplus3T/6.0.1/rootdir/etc/init.qcom.rc#L741

Most Android kernels do come with module loading enabled, but OEMs don't automatically load the kernel modules they build on an as-needed basis. Instead of doing this, I disabled modules in my kernel and built in the drivers that would always be loaded anyway.

**Comment by zx2c4** [comment]
Ah, that's too bad. In that case, then, would you be willing to at least _allow_ module loading, while still compiling everything you desire in? That way I can distribute the WireGuard module for your kernel. Alternatively, I can just distribute my own kernels, which might be easier anyway.

I'll go through your default config and disable a lot of the junk that nobody wants and increases attack surface (NetBIOS connection tracking? No thank you...), and I'll submit a separate pull request for that. Seems like a worthy mission.

**Comment by kerneltoast** [comment]
Are you planning on making a module that'll work with the kernel binary distributed in my ROM or are you planning on compiling my kernel yourself?

Looks like that NetBIOS stuff is part of the base Android config: https://github.com/android/kernel_common/blob/android-3.18/android/configs/android-base.cfg#L120

I really don't know anything about the networking drivers, so I'm not comfortable with disabling some of them without making sure that they aren't used for something important.

**Comment by zx2c4** [comment]
> Are you planning on making a module that'll work with the kernel binary distributed in my ROM or are you planning on compiling my kernel yourself?

I was planning on building an external module against your kernel sources. If you decide not to enable modules, I was planning on just distributing my own kernel based on yours using [this clever and obscene trick](https://paste.pound-python.org/show/4Qg6thdQEVaXhUYo0nGD/).

> Looks like that NetBIOS stuff is part of the base Android config

Lots of things are in that base config that really shouldn't there, and luckily the fact that they're just unloaded modules keeps it out of the kernel.

> I really don't know anything about the networking drivers

I do. I'll write justifications in any pull requests I submit.

**Comment by zx2c4** [comment]
>> Are you planning on making a module that'll work with the kernel binary distributed in my ROM or are you planning on compiling my kernel yourself?

> I was planning on building an external module against your kernel sources. If you decide not to enable modules, I was planning on just distributing my own kernel based on yours using [this clever and obscene trick](https://paste.pound-python.org/show/4Qg6thdQEVaXhUYo0nGD/).

I think I'll make things less of a PITA for you and just distribute modified whole kernels instead.

**Comment by zx2c4** [comment]
As a matter of historical record, this PR has been entirely superseded by https://github.com/sultanxda/android/pull/7 .
